sample_consensus/include/pcl/sample_consensus/impl/sac_model_ellipse3d.hpp

```diff
@@ -0,0 +1,894 @@
+  Eigen::Vector3f ellipse_normal = common_helper_vec.normalized();
+
+  // The same check is implemented in isSampleGood, so be sure to look there too
+  // if you find the need to change something here.
+  if (common_helper_vec.squaredNorm() < Eigen::NumTraits<float>::dummy_precision ())
+  {
+    PCL_ERROR ("[pcl::SampleConsensusModelEllipse3D::computeModelCoefficients] Sample points too similar or collinear!\n");
+    return (false);
+  }
+
+  // Definition of the local reference frame of the ellipse
+  Eigen::Vector3f x_axis = (p1 - p0).normalized();
+  Eigen::Vector3f z_axis = ellipse_normal.normalized();
+  Eigen::Vector3f y_axis = z_axis.cross(x_axis).normalized();
+  
+  // Create the transposed rotation matrix
+  Eigen::Matrix3f Rot;
+  Rot << x_axis(0), y_axis(0), z_axis(0),
+      x_axis(1), y_axis(1), z_axis(1),
+      x_axis(2), y_axis(2), z_axis(2);
+
+  // Convert the points to local coordinates
+  Eigen::Vector3f p0_, p1_, p2_, p3_, p4_, p5_;
+  p0_ = Rot.transpose() * (p0 - p0);
+  p1_ = Rot.transpose() * (p1 - p0);
+  p2_ = Rot.transpose() * (p2 - p0);
+  p3_ = Rot.transpose() * (p3 - p0);
+  p4_ = Rot.transpose() * (p4 - p0);
+  p5_ = Rot.transpose() * (p5 - p0);
+  
+
+  // Fit an ellipse to the samples to obtain its conic equation parameters
+  // (this implementation follows the manuscript "Direct Least Square Fitting of Ellipses"
+  //  A. Fitzgibbon, M. Pilu and R. Fisher, IEEE TPAMI, 21(5) : 476–480, May 1999).
+
+  // xOy projections only
+  Eigen::VectorXf X(6);
+  X << p0_(0), p1_(0), p2_(0), p3_(0), p4_(0), p5_(0);
+  Eigen::VectorXf Y(6);
+  Y << p0_(1), p1_(1), p2_(1), p3_(1), p4_(1), p5_(1);
+
+  // Design matrix D
+  Eigen::MatrixXf D(6, 6);
+  D << X(0) * X(0), X(0) * Y(0), Y(0) * Y(0), X(0), Y(0), 1.0,
+      X(1) * X(1), X(1) * Y(1), Y(1) * Y(1), X(1), Y(1), 1.0,
+      X(2) * X(2), X(2) * Y(2), Y(2) * Y(2), X(2), Y(2), 1.0,
+      X(3) * X(3), X(3) * Y(3), Y(3) * Y(3), X(3), Y(3), 1.0,
+      X(4) * X(4), X(4) * Y(4), Y(4) * Y(4), X(4), Y(4), 1.0,
+      X(5) * X(5), X(5) * Y(5), Y(5) * Y(5), X(5), Y(5), 1.0;
+
+  // Scatter matrix S
+  Eigen::MatrixXf S = D.transpose() * D;
+
+  // Constraint matrix C
+  Eigen::MatrixXf C = Eigen::MatrixXf::Zero(6, 6);
+  C(0, 2) = -2.0;
+  C(1, 1) = 1.0;
+  C(2, 0) = -2.0;
+
+  // Solve the Generalized Eigensystem: S*a = lambda*C*a
+  Eigen::GeneralizedEigenSolver<Eigen::MatrixXf> solver;
+  solver.compute(S, C);
+  Eigen::VectorXf eigvals = solver.eigenvalues().real();
+
+  // Find the negative eigenvalue 'neigvec' (the largest, if many exist)
+  int idx(-1);
+  float absmin(0.0);
+  for (size_t i(0); i < (size_t) eigvals.size(); ++i) {
+    if (eigvals(i) < absmin && !std::isinf(eigvals(i))) {
+      idx = i;
+    }
+  }
+  // Return "false" in case the negative eigenvalue was not found
+  if (idx == -1) {
+    PCL_ERROR("[pcl::SampleConsensusModelEllipse3D::computeModelCoefficients] Failed to find the negative eigenvalue in the GES.\n");
```

Comment:
I am not sure what this means? This appears many times in the [test log](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=22222&view=logs&j=b29c579b-56da-5a9f-b5ad-e9e014f41171&t=3589a9c7-0f5f-5904-777a-0430ef177198&l=6843). Does this happen when the points don't form an ellipse in 2D? When there is an outlier among the 6 points? How about changing this to `PCL_DEBUG` (which is not displayed unless the user requests it)?

Reply:
Yes, this happens when the solver is not able to find the solution for those 6 points, meaning that the problem is ill-posed - one or more points suffer from a large deviation, even though the authors claim that the method can handle noisy data. I guess that the best approach would be to change it to `PCL_DEBUG` as you suggested, because this is likely to occur a considerable amount of times during RANSAC fitting.